src/server/pkg/storage/fileset/pg_store.go

```diff
@@ -0,0 +1,108 @@
+package fileset
+
+import (
+	"context"
+	"database/sql"
+	"testing"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/jmoiron/sqlx"
+	"github.com/pachyderm/pachyderm/src/server/pkg/dbutil"
+	"github.com/pachyderm/pachyderm/src/server/pkg/storage/fileset/index"
+)
+
+var _ Store = &pgStore{}
+
+type pgStore struct {
+	db *sqlx.DB
+}
+
+func NewPGStore(db *sqlx.DB) Store {
+	return &pgStore{db: db}
+}
+
+func (s *pgStore) PutIndex(ctx context.Context, p string, idx *index.Index) error {
+	if idx == nil {
+		idx = &index.Index{}
+	}
+	data, err := proto.Marshal(idx)
+	if err != nil {
+		return err
+	}
+	_, err = s.db.ExecContext(ctx,
+		`INSERT INTO storage.paths (path, index_pb)
+		VALUES ($1, $2)
+		ON CONFLICT (path) DO UPDATE SET index_pb = $2
+		`, p, data)
+	return err
+}
+
+func (s *pgStore) GetIndex(ctx context.Context, p string) (*index.Index, error) {
+	var indexData []byte
+	if err := s.db.GetContext(ctx, &indexData, `SELECT index_pb FROM storage.paths WHERE path = $1`, p); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, ErrPathNotExists
+		}
+		return nil, err
+	}
+	idx := &index.Index{}
+	if err := proto.Unmarshal(indexData, idx); err != nil {
+		return nil, err
+	}
+	return idx, nil
+}
+
+func (s *pgStore) Walk(ctx context.Context, prefix string, cb func(string) error) error {
+	rows, err := s.db.QueryContext(ctx, `SELECT path from storage.paths WHERE path LIKE $1 || '%'`, prefix)
+	if err != nil {
+		return err
+	}
+	defer rows.Close()
```

Comment:
We should probably capture any errors here.

Reply:
Hmm, it looks like you might not have to call `Close` based on the documentation, but it doesn't seem clear to me what happens when errors are encountered. It might be safer to have something like this:
```
defer func() {
  if err := rows.Close(); retErr == nil {
    retErr = err
  }
  if err := rows.Err(); retErr == nil {
    retErr = err
  }
}
```

Reply:
I can't be the first person to think of this, but this is great
```go
defer errutil.SetRetErrIfNil(&retErr, rows.Close())
```

Reply:
might need to wrap in function, but still it's very convenient.

Reply:
yeah you have to 
```go
defer func() { errutil.SetRetErrIfNil(&retErr, rows.Close()) }()
```
to prevent the arguments from being evaluated

Reply:
I am a bit concerned about this negatively impacting readability. Lets not do this for the time being.